### ISSUE TITLE
GCC 11 build errors

### ISSUE THREAD

**Issue by susnux** [oneapi-src/oneDNN#1075]
# Summary
Building with gcc 11 results in build issues:
in `src/common/primitive_cache.cpp:17`
  * `no known conversion for argument from 'const std::thread::id' to 'const std::error_condition`
  * `no known conversion for argument 1 from 'const std::thread::id' to 'const memory_desc_t&`

# Version
Version 2.2.3 from github releases section

# Environment
* CPU: Different x86_64
* OS version: openSUSE Factory (Linux 5.12.4)
* Compiler version: GCC 11.1.1
* CMake version: 3.20.2

# Steps to reproduce
Build with gcc 11

# Observed behavior
Here is the log:
```
 In file included from /home/abuild/rpmbuild/BUILD/oneDNN-2.2.3/src/common/primitive_attr.hpp:27,
                 from /home/abuild/rpmbuild/BUILD/oneDNN-2.2.3/src/common/primitive_hashing.hpp:25,
                 from /home/abuild/rpmbuild/BUILD/oneDNN-2.2.3/src/common/primitive_cache.hpp:27,
                  from /home/abuild/rpmbuild/BUILD/oneDNN-2.2.3/src/common/primitive_cache.cpp:17:
 /home/abuild/rpmbuild/BUILD/oneDNN-2.2.3/src/common/type_helpers.hpp:314:13: note: candidate: 'bool dnnl::impl::operator!=(const memory_desc_t&, const memory_desc_t&)'
   314 | inline bool operator!=(const memory_desc_t &lhs, const memory_desc_t &rhs) {
       |             ^~~~~~~~
 /home/abuild/rpmbuild/BUILD/oneDNN-2.2.3/src/common/type_helpers.hpp:314:45: note:   no known conversion for argument 1 from 'const std::thread::id' to 'const memory_desc_t&' {aka 'const dnnl_memory_desc_t&'}
   314 | inline bool operator!=(const memory_desc_t &lhs, const memory_desc_t &rhs) {
       |                        ~~~~~~~~~~~~~~~~~~~~~^~~
 In file included from /usr/include/c++/11/mutex:42,
                  from /usr/include/c++/11/future:38,
                  from /home/abuild/rpmbuild/BUILD/oneDNN-2.2.3/src/common/primitive_cache.hpp:20,
                  from /home/abuild/rpmbuild/BUILD/oneDNN-2.2.3/src/common/primitive_cache.cpp:17:
 /usr/include/c++/11/system_error:400:3: note: candidate: 'bool std::operator!=(const std::error_code&, const std::error_code&)'
   400 |   operator!=(const error_code& __lhs, const error_code& __rhs) noexcept
       |   ^~~~~~~~
 /usr/include/c++/11/system_error:400:32: note:   no known conversion for argument 1 from 'const std::thread::id' to 'const std::error_code&'
   400 |   operator!=(const error_code& __lhs, const error_code& __rhs) noexcept
       |              ~~~~~~~~~~~~~~~~~~^~~~~
 /usr/include/c++/11/system_error:406:3: note: candidate: 'bool std::operator!=(const std::error_code&, const std::error_condition&)'
   406 |   operator!=(const error_code& __lhs, const error_condition& __rhs) noexcept
       |   ^~~~~~~~
 /usr/include/c++/11/system_error:406:32: note:   no known conversion for argument 1 from 'const std::thread::id' to 'const std::error_code&'
   406 |   operator!=(const error_code& __lhs, const error_condition& __rhs) noexcept
       |              ~~~~~~~~~~~~~~~~~~^~~~~
 /usr/include/c++/11/system_error:412:3: note: candidate: 'bool std::operator!=(const std::error_condition&, const std::error_code&)'
   412 |   operator!=(const error_condition& __lhs, const error_code& __rhs) noexcept
       |   ^~~~~~~~
 /usr/include/c++/11/system_error:412:37: note:   no known conversion for argument 1 from 'const std::thread::id' to 'const std::error_condition&'
   412 |   operator!=(const error_condition& __lhs, const error_code& __rhs) noexcept
       |              ~~~~~~~~~~~~~~~~~~~~~~~^~~~~
 /usr/include/c++/11/system_error:417:3: note: candidate: 'bool std::operator!=(const std::error_condition&, const std::error_condition&)'
   417 |   operator!=(const error_condition& __lhs,
       |   ^~~~~~~~
 /usr/include/c++/11/system_error:417:37: note:   no known conversion for argument 1 from 'const std::thread::id' to 'const std::error_condition&'
   417 |   operator!=(const error_condition& __lhs,
       |              ~~~~~~~~~~~~~~~~~~~~~~~^~~~~
 make[2]: *** [src/common/CMakeFiles/dnnl_common.dir/build.make:457: src/common/CMakeFiles/dnnl_common.dir/primitive_cache.cpp.o] Error 1
```

# Expected behavior
No errors

**Comment by vpirogov** [comment]
Thank you for the report, @susnux. I confirm compilation issues with GCC 10 and GCC 11.

**Comment by densamoilov** [comment]
Hi @susnux, the issues have been fixed in master in 1fe0f2594a1bfc6386fd8f6537f971d5ae9c1214. See #1023 for more details. Even though the build issues are fixed now you will continue observing a bunch of warnings. We are aware of those warnings and will work to address them.

**Comment by susnux** [comment]
> Hi @susnux, the issues have been fixed in master in [1fe0f25](https://github.com/oneapi-src/oneDNN/commit/1fe0f2594a1bfc6386fd8f6537f971d5ae9c1214).

I can confirm, cherry-picking this commit on top of tag 2.2.3 does solve the issue!

**Comment by vpirogov** [comment]
Fix is backported to oneDNN v2.2.4.
